### PR TITLE
Update css-split-webpack-plugin to listen for webpack emit

### PR DESCRIPTION
**Pull request by didi0613**
Issue:
For IE9, css splitting before optimization, which causing an incorrect z-index reduction under production.

Proposal:
- keep the original logic
- add an `listenForEmit` option to the plugin to listen for webpack emit
The PR added an option to the plug-in to run on emit, so it maintain backwards compatibility

### CODE REVIEW

**Comment by izaakschroeder** [comment]
See if you can get CI passing here @didi0613?

**Comment by izaakschroeder** [comment]
I think this is less than ideal but I understand the pragmatic reasons for having it. Seems fine to me, maybe a few tweaks.

File: src/index.js [hunk 1]
```diff
@@ -77,12 +77,14 @@ export default class CSSSplitWebpackPlugin {
     imports = false,
     filename = '[name]-[part].[ext]',
     preserve,
+    listenForEmit = false
```
Comment (izaakschroeder) [inline src/index.js]: I think this name is a bit misleading? Maybe we can do like the last call plugin does and add `phase` or something like that which defaults to `optimize-chunk-assets`. Documentation for this should also be added to the README.

File: src/index.js [hunk 2]
```diff
@@ -114,12 +116,49 @@ export default class CSSSplitWebpackPlugin {
...
             name(i),
-            map.toString()
+            map && map.toString()
           );
         }),
       });
     });
   }
+  
+  chunksMapping(chunks, assets, publicPath) {
```
Comment (izaakschroeder) [inline src/index.js]: You can add `done` to the list of args to and DRY up some duplicate code you have in the below section:

```
Promise.all(promises).then(() => {
  done();
}, done);
```

Comment (izaakschroeder) [reply src/index.js]: You may also wish to pass `compilation` instead of `publicPath` as you have some duplicate `publicPath` code below as well.

**Comment by jchip** [comment]
@izaakschroeder another approach is do it like the CSS optimize plugin using the last call plugin, which listens to all compilation/optimize-chunks-assets event, but check for existence of CSS in assets.  The approach @didi0613 took here would allow a patch release.

File: src/index.js (hunk 1 above)
Comment (didi0613) [reply src/index.js]: @izaakschroeder Yea i agree.
Here, we are specifying compiler plugin, which is `emit` or `this-compilation` (as default). Currently only accept `emit` as an option. I've re-named as `compilerPhase`, please review.

**Comment by izaakschroeder** [comment]
I'm fine with this @jchip @didi0613. It's not ideal but it won't blow anything up and it obviously solves a real problem. I think CI was complaining about invalid peer dependency on a module that isn't even required. Should be fixed in latest `master`. You can rebase and squash and should be just about good to go.

**Comment by didi0613** [comment]
@izaakschroeder Checks have passed. Please let me know if anything else i need to fix. Thank you

**Comment by izaakschroeder** [comment]
I just cleaned a few things up in your branch here https://github.com/metalabdesign/css-split-webpack-plugin/pull/21. You're still listed as the author. 😄 Available now as `+ css-split-webpack-plugin@0.2.4`.

**Comment by izaakschroeder** [comment]
Thanks again for contributing @didi0613 @jchip 🎉

**Pull request closed without merging**